### PR TITLE
Make Cell & HeaderCell generic

### LINKED ISSUE #472: Table is generic, but Cell is not
Thanks for creating this library. The API and feature set are quite nice.  I have a minor suggestion regarding the use of generics in Table component.

### What problem does this feature solve?

Currently Table component is generic, but the Cell component is not generic. So if we have a usage like this: 

```
<Table<InventoryItem, string> data={inventoryItems}>
    <Column
        resizable
        width={200}
    >
        <HeaderCell>Produce Type</HeaderCell>
        <Cell>
            {(row) => row.produceType.name}
        </Cell>
    </Column>
   {/* ... */}
</Table>
```

`row.produceType` has `any` type because Cell component has no idea about the type parameter received by the Table component. This can be a subtle source of bugs because at a glance the code looks fully type safe. 

In fact actually adding a type annotation to row is also not enough because the default type used by Cell uses never. 

### What does the proposed API look like?

We could make the Cell component generic too and make the Table component accept a render prop which injects Cell with the correct type. 
 
```
<Table<InventoryItem, string> data={inventoryItems}>
    {({ Column, HeaderCell, Cell }) => (
       <>
        <Column
            resizable
            width={200}
        >
            <HeaderCell>Produce Type</HeaderCell>
            <Cell>
                {(row) => row.produceType.name}
            </Cell>
        </Column>
       {/* ... */}
      </>
  )}
</Table>
```
IMHO this will fit well with rest of the API. This can be implemented in backward compatible manner, so will not break existing code.

### PR DESCRIPTION
**Pull request by lorefnon**
Resolves #472

### CODE REVIEW

**Comment by simonguo** [comment]
Hi  @lorefnon  , tests failed. 
You can execute the test case locally with the following command.

```
npm run test
// or
npm run tdd
```

**Comment by lorefnon** [comment]
Hi @simonguo - I have addressed the failures.

File: test/Table.test.tsx [hunk 1]
```diff
@@ -61,6 +61,17 @@ const ref = React.createRef<TableInstance<Row, string>>();
...
+  {({ Column, HeaderCell, Cell }) => (
+    <>
+      <Column>
+        <HeaderCell>Id</HeaderCell>
+        <Cell>{({ id }) => id}</Cell>
+      </Column>
+    </>
+  )}
+</Table>;
+
```
Comment (simonguo) [inline test/Table.test.tsx:74]: It is recommended to add a type check.

```suggestion
interface InventoryItem {
  id: string;
  name: string;
}

const table = React.createRef<TableInstance<InventoryItem, string>>();

<Table<InventoryItem, string> ref={table}>
  {({ Column, HeaderCell, Cell }) => (
    <>
      <Column>
        <HeaderCell>Name</HeaderCell>
        <Cell>{row => row.name}</Cell>
      </Column>
      <Column>
        <HeaderCell>Type</HeaderCell>
        {/** @ts-expect-error Property 'type' does not exist on type 'InventoryItem' */}
        <Cell>{row => row.type}</Cell>
      </Column>
    </>
  )}
</Table>;
```

**Review by simonguo — CHANGES_REQUESTED** [review]
Hi @b1rdex, `children` prop should be added to `<Table>` in readme.

https://github.com/rsuite/rsuite-table/tree/main#table

**Review by simonguo — APPROVED** [review]
LGTM

**Pull request merged**